lxd/db_storage_pools.go

```diff
@@ -0,0 +1,496 @@
+	volumeID, _, err := dbStoragePoolVolumeGetType(db, volumeName, volumeType, poolID)
+	if err != nil {
+		return err
+	}
+
+	tx, err := dbBegin(db)
+	if err != nil {
+		return err
+	}
+
+	err = dbStorageVolumeConfigClear(tx, volumeID)
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+
+	err = dbStorageVolumeConfigAdd(tx, volumeID, volumeConfig)
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+
+	return txCommit(tx)
+}
+
+// Delete storage volume attached to a given storage pool.
+func dbStoragePoolVolumeDelete(db *sql.DB, volumeName string, volumeType int, poolID int64) error {
+	volumeID, _, err := dbStoragePoolVolumeGetType(db, volumeName, volumeType, poolID)
+	if err != nil {
+		return err
+	}
+
+	_, err = dbExec(db, "DELETE FROM storage_volumes WHERE id=?", volumeID)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Rename storage volume attached to a given storage pool.
+func dbStoragePoolVolumeRename(db *sql.DB, oldVolumeName string, newVolumeName string, volumeType int, poolID int64) error {
+	volumeID, _, err := dbStoragePoolVolumeGetType(db, oldVolumeName, volumeType, poolID)
+	if err != nil {
+		return err
+	}
+
+	tx, err := dbBegin(db)
+	if err != nil {
+		return err
+	}
+
+	_, err = tx.Exec("UPDATE storage_volumes SET name=? WHERE id=? AND type=?", newVolumeName, volumeID, volumeType)
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+
+	return txCommit(tx)
+}
+
+// Create new storage volume attached to a given storage pool.
+func dbStoragePoolVolumeCreate(db *sql.DB, volumeName string, volumeType int, poolID int64, volumeConfig map[string]string) (int64, error) {
+	tx, err := dbBegin(db)
+	if err != nil {
+		return -1, err
+	}
+
+	result, err := tx.Exec("INSERT INTO storage_volumes (storage_pool_id, type, name) VALUES (?, ?, ?)", poolID, volumeType, volumeName)
+	if err != nil {
+		tx.Rollback()
+		return -1, err
+	}
+
+	volumeID, err := result.LastInsertId()
+	if err != nil {
+		tx.Rollback()
+		return -1, err
+	}
+
+	err = dbStorageVolumeConfigAdd(tx, volumeID, volumeConfig)
+	if err != nil {
+		tx.Rollback()
+		return -1, err
+	}
+
+	err = txCommit(tx)
+	if err != nil {
+		return -1, err
+	}
+
+	return volumeID, nil
+}
+
+// Get ID of a storage volume on a given storage pool of a given storage volume
+// type.
+func dbStoragePoolVolumeGetTypeID(db *sql.DB, volumeName string, volumeType int, poolID int64) (int64, error) {
+	volumeID := int64(-1)
+	query := `SELECT storage_volumes.id
+FROM storage_volumes
+JOIN storage_pools
+ON storage_volumes.storage_pool_id = storage_pools.id
+WHERE storage_volumes.storage_pool_id=?
+AND storage_volumes.name=? AND storage_volumes.type=?`
+	inargs := []interface{}{poolID, volumeName, volumeType}
+	outargs := []interface{}{&volumeID}
+
+	err := dbQueryRowScan(db, query, inargs, outargs)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return -1, NoSuchObjectError
+		}
```

Comment:
Similar to comment above. SmartError should handled the ErrNoRows just fine, this is likely to just confused it instead.

Also in the case where you get an error other than sql.ErrNoRows, you're ignoring the error and just moving on.